File: shipbob_to_gorgias/typescript/test/webhook.test.ts
import assert from 'node:assert/strict'
import * as fs from 'node:fs'
import * as os from 'node:os'
import * as path from 'node:path'
import { afterEach, beforeEach, test } from 'node:test'

import * as webhook from '../src/webhook.js'
import {
    GORGIAS_SECRETS,
    makeOnholdEvent,
    makePandium,
    makeShipmentEvent,
    recordingGorgias,
    webhookTrigger,
} from './helpers.js'

let tmpDir: string

beforeEach(() => {
    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'sb2g-'))
})

afterEach(() => {
    fs.rmSync(tmpDir, { recursive: true, force: true })
})

async function run(triggers: any[], opts: { metadata?: any; existing?: string[] } = {}) {
    const gorgias = recordingGorgias(opts.existing ?? ['jane@example.com'])
    const pandium = makePandium({
        secrets: GORGIAS_SECRETS,
        runTriggers: triggers,
        metadata: opts.metadata,
        tmpDir,
        runMode: 'webhook',
    })
    const result = await webhook.run(pandium, { gorgias })
    return { result, gorgias }
}

test('run opens a ticket and returns only processed_events', async () => {
    const triggers = [webhookTrigger(tmpDir, makeShipmentEvent(456789), 't1')]
    const { result, gorgias } = await run(triggers)

    const ticket = gorgias.log.ticket[0]
    assert.deepEqual(ticket.customer, { id: 40 }) // linked to the found customer
    assert.deepEqual(ticket.tags, [{ name: 'shipbob-shipment' }, { name: 'shipbob-delivered' }])
    assert.match(ticket.messages[0].body_text, /is now Delivered/)
    assert.match(ticket.messages[0].body_text, /USPS 9400100000000000000000/)
    assert.deepEqual(Object.keys(result), ['processed_events']) // leaves cron's cursor keys alone
    assert.ok('456789:Delivered' in result.processed_events)
})

test('run dedupes duplicate deliveries within a batch and prunes stale processed_events', async () => {
    const now = new Date()
    const triggers = [
        webhookTrigger(tmpDir, makeShipmentEvent(456789), 't1'),
        webhookTrigger(tmpDir, makeShipmentEvent(456789), 't2'), // duplicate
    ]
    const metadata = {
        processed_events: {
            '456790:Delivered': now.toISOString(), // recent -> kept
            '999999:Delivered': new Date(now.getTime() - 45 * 60 * 1000).toISOString(), // >30 min -> pruned
        },
    }
    const { result, gorgias } = await run(triggers, { metadata })

    assert.equal(gorgias.log.ticket.length, 1) // one ticket despite the duplicate delivery
    assert.deepEqual(
        new Set(Object.keys(result.processed_events)),
        new Set(['456789:Delivered', '456790:Delivered'])
    )
})

test('run tickets every distinct shipment+status once, even redeliveries of the same status', async () => {
    const triggers = [
        webhookTrigger(tmpDir, makeShipmentEvent(1, 'OnHold'), 't1'),
        webhookTrigger(tmpDir, makeShipmentEvent(1, 'OnHold'), 't2'),
        webhookTrigger(tmpDir, makeShipmentEvent(1, 'Delivered'), 't3'),
    ]
    const { result, gorgias } = await run(triggers)

    assert.equal(gorgias.log.ticket.length, 2)
    assert.deepEqual(new Set(Object.keys(result.processed_events)), new Set(['1:OnHold', '1:Delivered']))
})

test('run creates the customer by external_id when the recipient has no email, and formats OnHold body correctly', async () => {
    const triggers = [webhookTrigger(tmpDir, makeOnholdEvent(), 't1')]
    const { result, gorgias } = await run(triggers)

    const created = gorgias.log.create[0]
    assert.ok(!('email' in created))
    // the synthetic key the cron flow uses too: name address1 city country
    assert.equal(created.external_id, 'Jane Buyer 100 Nowhere Blvd Gotham City US')

    const ticket = gorgias.log.ticket[0]
    assert.deepEqual(ticket.customer, { id: 1001 }) // the customer we just created
    const body = ticket.messages[0].body_text
    assert.match(body, /is now OnHold/)
    assert.match(body, /Reason: Invalid Address; Payment Failure/)
    assert.doesNotMatch(body, /Tracking:/) // OnHold shipments carry none
    assert.match(body, /4 x Pinnacle Shampoo \(PIN-100\)/)
    assert.deepEqual(Object.keys(result.processed_events), ['107414278:OnHold'])
})

File: shipbob_to_gorgias/ruby/test/webhook_test.rb
require_relative 'test_helper'
require_relative 'fakes'
require_relative '../webhook'
require 'tmpdir'

class WebhookTest < Minitest::Test
  def test_prune_drops_entries_older_than_window_keeps_recent
    now = Time.utc(2026, 7, 9, 12, 0, 0)
    processed = {
      '1:Delivered' => (now - (10 * 60)).strftime('%Y-%m-%dT%H:%M:%S+00:00'), # 10 min ago -- kept
      '2:Delivered' => (now - (45 * 60)).strftime('%Y-%m-%dT%H:%M:%S+00:00'), # 45 min ago -- dropped
      '3:Delivered' => 'not-a-timestamp', # unparseable -- treated as expired, dropped
    }
    kept = Sb2Gorgias::Webhook.prune(processed, now)
    assert_equal ['1:Delivered'], kept.keys
  end

  def test_shipment_id_falls_through_on_literal_zero
    event = { 'id' => 0, 'shipment_id' => 42 }
    assert_equal '42', Sb2Gorgias::Webhook.shipment_id(event)
  end

  def test_shipment_id_prefers_id_when_present
    assert_equal '5', Sb2Gorgias::Webhook.shipment_id({ 'id' => 5, 'shipment_id' => 42 })
  end

  def test_build_ticket_delivered_event_html_and_text
    event = make_shipment_event
    ticket = Sb2Gorgias::Webhook.build_ticket(event, { 'id' => 1 })
    message = ticket['messages'][0]

    assert_includes message['body_text'], 'Tracking: USPS 9400100000000000000000'
    assert_includes message['body_text'], 'Delivered on: 2026-07-09'
    assert_includes message['body_text'], 'Items:'
    assert_includes message['body_html'], '<p><b>Tracking:</b> USPS 9400100000000000000000</p>'
    refute_includes message['body_html'], 'Delivered on' # HTML never gets this block
    assert_equal [{ 'name' => 'shipbob-shipment' }, { 'name' => 'shipbob-delivered' }], ticket['tags']
  end

  def test_build_ticket_onhold_event_no_tracking_no_email_reason_from_description
    event = make_onhold_event
    ticket = Sb2Gorgias::Webhook.build_ticket(event, { 'id' => 1, 'external_id' => 'Jane Buyer 100 Nowhere Blvd Gotham City US' })
    message = ticket['messages'][0]

    refute_includes message['body_text'], 'Tracking:'
    assert_includes message['body_text'], 'Reason: Invalid Address; Payment Failure'
    assert_equal [{ 'name' => 'shipbob-shipment' }, { 'name' => 'shipbob-onhold' }], ticket['tags']
  end

  def test_run_dedupes_within_a_single_batch
    Dir.mktmpdir do |dir|
      event = make_shipment_event(1)
      triggers = [webhook_trigger(dir, event, 't1'), webhook_trigger(dir, event, 't2')]
      pandium = make_pandium(secrets: GORGIAS_SECRETS, run_triggers: triggers)
      gorgias = recording_gorgias

      result = Sb2Gorgias::Webhook.run(pandium, gorgias: gorgias, now: Time.now.utc)

      assert_equal 1, gorgias.log[:ticket].size
      assert_equal 1, result['processed_events'].size
    end
  end

  def test_run_creates_separate_tickets_for_different_statuses
    Dir.mktmpdir do |dir|
      delivered = make_shipment_event(1, 'Delivered')
      onhold = make_shipment_event(1, 'OnHold')
      triggers = [webhook_trigger(dir, delivered, 't1'), webhook_trigger(dir, onhold, 't2')]
      pandium = make_pandium(secrets: GORGIAS_SECRETS, run_triggers: triggers)

      result = Sb2Gorgias::Webhook.run(pandium, gorgias: recording_gorgias, now: Time.now.utc)

      assert_equal %w[1:Delivered 1:OnHold].sort, result['processed_events'].keys.sort
    end
  end

  def test_run_leaves_event_unprocessed_when_ticket_creation_fails
    Dir.mktmpdir do |dir|
      event = make_shipment_event(1)
      triggers = [webhook_trigger(dir, event, 't1')]
      pandium = make_pandium(secrets: GORGIAS_SECRETS, run_triggers: triggers)
      gorgias = recording_gorgias
      gorgias.define_singleton_method(:create_ticket) { |_payload| raise 'boom' }

      result = Sb2Gorgias::Webhook.run(pandium, gorgias: gorgias, now: Time.now.utc)

      assert_empty result['processed_events']
    end
  end

  def test_run_result_has_only_processed_events_key
    Dir.mktmpdir do |dir|
      pandium = make_pandium(secrets: GORGIAS_SECRETS, run_triggers: [])
      result = Sb2Gorgias::Webhook.run(pandium, gorgias: recording_gorgias, now: Time.now.utc)
      assert_equal ['processed_events'], result.keys
    end
  end
end
